Extend SDK layer 2 docs (#4980)
opencv/cvat: site/content/en/docs/api_sdk/sdk/highlevel-api.md

## Overview

This layer provides high-level APIs, allowing easier access to server operations.

> API includes _Repositories_ and _Entities_. _Repositories_ provide management
> operations for _Entities_. _Entities_ represent objects on the server

Reviewer: I would suggest only using italics for the defining occurrences of the terms (once for each term). Otherwise it's going to be annoying to consistently italicize these terms everywhere - in fact, I already see them appear without italics at the bottom of the document. I don't think italics everywhere would help with readability either.

```suggestion
API includes Repositories and Entities. _Repositories_ provide management
operations for Entities. _Entities_ represent objects on the server
```

Author: Agree. I decided to leave them in 2 places, where the terms can be declared for a reader.

This layer provides high-level APIs, allowing easier access to server operations.
API includes _Repositories_ and _Entities_. Repositories provide management
operations for Entitites. Entitites represent separate objects on the server

> (e.g. tasks, jobs etc). The key difference from the low-level API is that

Reviewer: ```suggestion
(e.g. tasks, jobs etc). A key difference from the low-level API is that
```

I don't think you could call it _the_ key difference when there are other important differences as well (like representing operations as methods rather than standalone functions).

Author: Class methods are not a big improvement over free functions, especially when we speak about this first layer, where apis and models are already grouped into classes. It's not that important in which order we put words, dots and dashes in the function call (`r = p_cls_foo(p, q)` vs `r = p.foo(q)` vs even `p_cls_foo(p, q, r)`). The one real benefit of this is the ability to inherit and overload, but its not something that I'd like to talk about in the docs. I've changed the wording.

The `Client` class implements the [context manager protocol](https://docs.python.org/3/reference/datamodel.html#context-managers).
When the context is closed, the session is finished, and the user is logged out
automatically. Otherwise, these actions can be done with the `close()` and `logout()` methods.

You can create and start using a `Client` instance this way:

```python
from cvat_sdk import make_client

with make_client('localhost', port='8080', credentials=('user', 'password')) as client:
    ...
```

The `make_client()` function handles configuration and object creation for you.
It also allows to authenticate right after the object is created.

If you need to configure `Client` parameters, you can do this:

```python
from cvat_sdk import Config, Client

config = Config()
# set up some config fields ...

with Client('localhost:8080', config=config) as client:
    client.login(('user', 'password'))
    ...
```

You can specify server address both with and without the scheme. If the scheme is omitted,
it will be deduced automatically.

> The checks are performed in the following
order: `https` (with the default port 8080), `http` (with the default port 80).

> In some cases it may lead to incorrect results - eg. you have 2 servers running on the

Reviewer: ```suggestion
In some cases it may lead to incorrect results - e.g. you have 2 servers running on the
```

same host at default ports. In such cases just specify the schema manually: `https://localhost`.

When the server is located, its version is checked. If an unsupported version is found,
an error can be raised or suppressed (controlled by `config.allow_unsupported_server`).
If the error is suppressed, some SDK functions may not work as expected with this server.
By default, a warning is raised and the error is suppressed.

> Please note that all `Client` operations rely on the server API and depend on the current user
rights. This affects the set of available APIs, objects and actions. For example, a regular user
can only see and modify their tasks and jobs, while an admin user can see all the tasks etc.

## Entities and Repositories

_Entities_ represent objects on the server. They provide read access to object fields
and implement additional relevant operations, including both the general Read-Update-Delete and
object-specific ones. The set of available general operations depends on the object type.

_Repositories_ provide management operations for corresponding _Entities_. Typically, you don't

> need to create _Repository_ objects manually. To obtain a _Repository_ object, use the

Reviewer: Do you _ever_ need to create Repository objects manually? Seems like the "typically" could be removed.

Author: Currently the answer is no, it's never needed. But it wasn't done on purpose.

    task.update({'name': 'mytask'})
    task.remove()
```

## Client

> The `cvat_sdk.core.client.Client` class manages session, provides connection management and

Reviewer: ```suggestion
The `cvat_sdk.core.client.Client` class manages sessions, provides connection management and
```

Author: Changed the wording

> resource location operations. It is the staring point for using CVAT SDK.

Reviewer: ```suggestion
resource location operations. It is the starting point for using CVAT SDK.
```

An instance of `Client` can be created directly by calling the class constructor,
or with the utility function `cvat_sdk.core.client.make_client()`, which can handle
some configuration for you. The extended configuration for `Client` is performed with
the `cvat_sdk.core.client.Config` class instances. A `Config` object can be passed to

> the `Client` constructor and then it can be accessed with the `Client.config` field.

Reviewer: ```suggestion
the `Client` constructor and then it can be accessed via the `Client.config` field.
```

Author: Changed the wording

> `Client` objects implement [context manager protocol](https://docs.python.org/3/reference/datamodel.html#context-managers).

Reviewer: ```suggestion
`Client` objects implement the [context manager protocol](https://docs.python.org/3/reference/datamodel.html#context-managers).
```

This could also use a description of _how_ these objects implement the protocol, i.e., what happens on exit.

Author: Added

> You can create `Client` this way:

Reviewer: ```suggestion
You can create a `Client` instance this way:
```

```python
from cvat_sdk import make_client

with make_client('localhost', port='8080', credentials=(user, password)) as client:
    ...
```

> 

Reviewer: It would be useful here to provide an equivalent example using `Client`, so that the reader can understand what `make_client` does for you.

Author: I've added some explanations. I don't think it is a good idea to fix the actions done by this function in the documentation, because the code tends to change quickly.

> Or, if you need to tweak `Client`, you can do this:

Reviewer: ```suggestion
Or, if you need to tweak the created `Client`, you can do this:
```

Author: Changed the wording

```python
from cvat_sdk import Config, Client

config = Config()
# set up some config fields ...

with Client('localhost:8080', config=config) as client:
    ...
```

> With `Client`, you can `login()` and `logout()`, and you can get Repository objects.

Reviewer: ```suggestion
With a `Client` instance, you can call `login()` and `logout()`, and you can get Repository objects.
```

Author: Changed the wording

## Entities and Repositories

> Entitites represent separate objects on the server. They provide read access to object fields

Reviewer: ```suggestion
Entities represent distinct objects on the server. They provide read access to object fields
```

Author: Changed the wording

and provide additional operations, including both general Read-Update-Delete and
object-specific ones.

Repositories provide management operations for corresponding Entities. Typically, you don't
need to create Repository objects manually. To obtain a Repository object, use the corresponding
`Client` instance property:

```python
client.projects
client.tasks
client.jobs
client.users
...
```

An entity can be created on the server with the corresponding Repository method `create()`:

```python
task = client.tasks.create(<task config>)
```

We can retrieve server objects using the `retrieve()` method of the Repository:

```python
job = client.jobs.retrieve(<job id>)
```

After calling these functions, we obtain local objects, representing their server counterparts.

Object fields can be updated with the `update()` method:

```python
job.update({'stage': 'validation'})
```

> The sever object will be updated and the local object will reflect the latest object state

Reviewer: ```suggestion
The server object will be updated and the local object will reflect the latest object state
```

after calling this operaiton.

> Note, that local objects may out of sync with their server counterparts for different reasons.

Reviewer: ```suggestion
Note that local objects may fall out of sync with their server counterparts for different reasons.
```

If you need to update the local object with the latest server state, use the `fetch()` method:

```python
# obtain 2 local copies of the same job
job_ref1 = client.jobs.retrieve(1)
job_ref2 = client.jobs.retrieve(1)

# update the server object with the first reference
job_ref1.update(...)
# job_ref2 is outdated now

job_ref2.fetch()
# job_ref2 is synced
```

> The local changes in the object are discarded when `fetch()` is called.

Reviewer: This seems a bit contradictory, since:

1. You haven't explained how to make local changes, and
2. Earlier you said that objects provide read-only fields, which weakly implies that local changes aren't possible.

Author: Changed the wording

Reviewer: You didn't address the main issue here - it's still unclear how local changes might appear.

Author: I don't know. maybe the user changed some values occasionally, because full immutability is impossible in python. Maybe the user created the entity object manually from the response model data. Previously, there was a holistic idea of changing fields and sending updates separately, but currently it works bad, because response and request model fields doesn't map directly. From my point, "any" is good enough to cover these cases.

Reviewer: We need to be consistent when describing how the user can interact with the API. If we mention local changes here, then that implies that it's possible to make local changes, and we need to describe that. Conversely, if it's _not_ possible to make local changes, then they should not be mentioned.

Note that by "possible" I mean "possible via supported, documented functionality". It goes without saying that in Python, you can pretty much modify anything if you're willing to go far enough. But the purpose of the documentation is to describe how users _should_ interact with the API, and so it only needs to cover situations that may arise from such legitimate use.

> Finally, if you need to remove object from the server, you can use the `remove()` method.

Reviewer: ```suggestion
Finally, if you need to remove the object from the server, you can use the `remove()` method.
```

Also, whose method?

This should probably be supplemented with an example.

Author: Added an example
